engine/src/build/toolchain/apple/swiftc.py

```diff
@@ -0,0 +1,649 @@
+        old_content = stream.read()
+
+      if old_content == new_content:
+        return
+
+    with open(self._filepath, 'w', encoding=self._encoding) as stream:
+      stream.write(new_content)
+
+  def write(self, data):
+    self._stringio.write(data)
+
+
+@contextlib.contextmanager
+def existing_directory(path):
+  """Returns a context manager wrapping an existing directory."""
+  yield path
+
+
+def create_stamp_file(path):
+  """Writes an empty stamp file at path."""
+  with FileWriter(path) as stream:
+    stream.write('')
+
+
+def create_build_cache_dir(args, build_signature):
+  """Creates the build cache directory according to `args`.
+
+  This function returns an object that implements the context manager
+  protocol and thus can be used in a with-clause. If -derived-data-dir
+  argument is not used, the returned directory is a temporary directory
+  that will be deleted when the with-clause is exited.
+  """
+  if not args.derived_data_dir:
+    return tempfile.TemporaryDirectory()
+
+  # The derived data cache can be quite large, so delete any obsolete
+  # files or directories.
+  stamp_name = f'{args.module_name}.stamp'
+  if os.path.isdir(args.derived_data_dir):
+    for name in os.listdir(args.derived_data_dir):
+      if name not in (build_signature, stamp_name):
+        path = os.path.join(args.derived_data_dir, name)
+        if os.path.isdir(path):
+          shutil.rmtree(path)
+        else:
+          os.unlink(path)
+
+  ensure_directory(args.derived_data_dir)
+  create_stamp_file(os.path.join(args.derived_data_dir, stamp_name))
+
+  return existing_directory(
+      ensure_directory(os.path.join(args.derived_data_dir, build_signature)))
+
+
+def ensure_directory(path):
+  """Creates directory at `path` if it does not exists."""
+  if not os.path.isdir(path):
+    os.makedirs(path)
+  return path
+
+
+def build_signature(env, args):
+  """Generates the build signature from `env` and `args`.
+
+  This allow re-using the derived data dir between builds while still
+  forcing the data to be recreated from scratch in case of significant
+  changes to the build settings (different arguments or tool versions).
+  """
+  m = hashlib.sha1()
+  for key in sorted(env):
+    if key.endswith('_VERSION') or key == 'DEVELOPER_DIR':
+      m.update(f'{key}={env[key]}'.encode('utf8'))
+  for i, arg in enumerate(args):
+    m.update(f'{i}={arg}'.encode('utf8'))
+  return m.hexdigest()
+
+
+def generate_source_output_file_map_fragment(args, filename):
+  """Generates source OutputFileMap.json fragment according to `args`.
+
+  Create the fragment for a single .swift source file for OutputFileMap.
+  The output depends on whether -whole-module-optimization argument is
+  used or not.
+  """
+  assert os.path.splitext(filename)[1] == '.swift', filename
+  basename = os.path.splitext(os.path.basename(filename))[0]
+  rel_name = os.path.join(args.target_out_dir, basename)
```

Comment:
Wasn't clear to me at first

```suggestion
  relative_name = os.path.join(args.target_out_dir, basename)
```